File: skills/ecosystem-services-assessment/scripts/tradeoff_analysis.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Usage: Rscript tradeoff_analysis.R <es_summary_table.csv> <output_dir>
# ES trade-off and synergy analysis across pixels or land cover units
# Usage: Rscript tradeoff_analysis.R <es_summary_csv> <output_dir>
# Requires: dplyr, ggplot2, corrplot, tidyr

# ── Inline logger ─────────────────────────────────────────────────────────────
SKILL_NAME <- "ecosystem-services-assessment"
.log_ts  <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
log_info <- function(...) message(.log_ts(), " [INFO]  ", sprintf(...))
log_warn <- function(...) message(.log_ts(), " [WARN]  ", sprintf(...))
log_error<- function(...) message(.log_ts(), " [ERROR] ", sprintf(...))
log_step <- function(n, d) log_info("-- STEP %d: %s", n, d)
log_decision <- function(v, val, why) log_info("DECISION | %s = %s | %s", v, val, why)
dir.create("logs", recursive=TRUE, showWarnings=FALSE)

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(corrplot)
  library(tidyr)
})

args       <- commandArgs(trailingOnly = TRUE)
es_file    <- ifelse(length(args) >= 1, args[1], "outputs/ecosystem_services/es_summary_table.csv")
output_dir <- ifelse(length(args) >= 2, args[2], "outputs/ecosystem_services")
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

log_info("Skill: %s | es_file=%s | output_dir=%s", SKILL_NAME, es_file, output_dir)

# ── Input precondition check ──────────────────────────────────────────────────
if (!file.exists(es_file)) {
  log_error(
    "Input not found: %s\nProbable cause: o script de quantificacao de servicos ecossistemicos nao foi executado ou o caminho esta errado.\nCheck: execute primeiro o script de mapeamento/quantificacao de ES.\nPrevious skill: ecosystem-services-assessment (quantification step).",
    es_file
  )
  stop("Missing: ", es_file)
}

log_step(1, "Load ecosystem services table")
es <- tryCatch({
  read.csv(es_file)
}, error = function(e) {
  log_error(
    "Failed to read CSV de servicos ecossistemicos: %s\nProbable cause: corrupted file ou com separador incorreto.\nCheck: abra o arquivo em editor de texto e confira o formato.\nPrevious skill: ecosystem-services-assessment (quantification step).",
    conditionMessage(e)
  )
  stop(e)
})

log_info("ES table loaded: %d land use classes", nrow(es))

n_na_total <- sum(is.na(es))
if (n_na_total > 0) {
  log_warn("ES table contains %d NA values total — correlations will be computed with 'complete.obs'.", n_na_total)
}

# ── Identify numeric ES columns ───────────────────────────────────────────────
log_step(2, "Identify ES indicator columns and normalise 0-1")
es_cols <- names(es)[sapply(es, is.numeric) & !names(es) %in% c("lulc_code", "n_pixels")]
log_info("ES indicators: %s", paste(es_cols, collapse = ", "))
log_decision("es_cols", paste(es_cols, collapse = ", "),
             "colunas numericas excluindo lulc_code e n_pixels sao tratadas como indicadores de ES")

if (length(es_cols) < 2) {
  log_warn("Fewer than 2 ES indicators found — trade-off analysis not possible with only %d column(s).", length(es_cols))
  log_info("Exiting without error. Add more indicators to the input CSV.")
  quit(status = 0)
}

# ── Normalise to 0-1 ──────────────────────────────────────────────────────────
es_norm <- tryCatch({
  es |>
    mutate(across(all_of(es_cols),
                  ~ (. - min(., na.rm=TRUE)) / (max(., na.rm=TRUE) - min(., na.rm=TRUE) + 1e-10)))
}, error = function(e) {
  log_error(
    "Failed to normalise indicators 0-1: %s\nProbable cause: non-numeric columns incorrectly identified as indicators.\nCheck: column types in the input CSV.\nPrevious skill: ecosystem-services-assessment (quantification step).",
    conditionMessage(e)
  )
  stop(e)
})

log_decision("normalization", "min-max [0,1] with epsilon 1e-10",
             "evita divisao por zero quando todos os valores de um indicador sao iguais")

# ── Correlation matrix (Spearman) ─────────────────────────────────────────────
log_step(3, "Compute Spearman correlation matrix among ES indicators")
cor_mat <- tryCatch({
  cor(es_norm[es_cols], method = "spearman", use = "complete.obs")
}, error = function(e) {
  log_error(
    "Failed to compute correlation matrix: %s\nProbable cause: all values in some column are NA after normalisation.\nCheck: presence of variation in ES indicators.\nPrevious skill: ecosystem-services-assessment (quantification step).",
    conditionMessage(e)
  )
  stop(e)
})

log_decision("correlation_method", "Spearman",
             "metodo nao-parametrico robusto a distribuicoes assimetricas comuns em dados de ES")

tryCatch({
  write.csv(as.data.frame(cor_mat), file.path(output_dir, "tradeoff_matrix.csv"))
  log_info("tradeoff_matrix.csv saved in: %s", output_dir)
}, error = function(e) {
  log_error(
    "Failed to salvar tradeoff_matrix.csv: %s\nProbable cause: permissao negada ou disco cheio.\nCheck: permissoes do output directory.\nPrevious skill: [none].",
    conditionMessage(e)
  )
  stop(e)
})

# ── Correlation heatmap ────────────────────────────────────────────────────────
log_step(4, "Generate trade-off heatmap (corrplot)")
tryCatch({
  png(file.path(output_dir, "tradeoff_heatmap.png"), width = 800, height = 700, res = 150)
  corrplot(cor_mat, method = "color", type = "upper", tl.cex = 0.8,
           addCoef.col = "black", number.cex = 0.7, cl.cex = 0.7,
           title = "ES Trade-offs (Spearman r)", mar = c(0, 0, 2, 0))
  dev.off()
  log_info("tradeoff_heatmap.png saved in: %s", output_dir)
}, error = function(e) {
  log_error(
    "Failed to generate heatmap de trade-offs: %s\nProbable cause: corrplot not installed ou matriz de correlacao invalida.\nCheck: se o pacote corrplot esta disponivel e a matriz tem pelo menos 2 variaveis.\nPrevious skill: [none].",
    conditionMessage(e)
  )
  stop(e)
})

# ── Scatter plots for top pairs ───────────────────────────────────────────────
log_step(5, "Generate scatter plots for ES indicator pairs")
pair_combos <- combn(es_cols, 2, simplify = FALSE)
n_pairs     <- min(6, length(pair_combos))
log_info("Generating %d scatter plots (of %d possible pairs)", n_pairs, length(pair_combos))
log_decision("max_scatter_plots", as.character(n_pairs),
             "limitado a 6 pares para evitar geracao excessiva de arquivos")

for (pr in pair_combos[seq_len(n_pairs)]) {
  tryCatch({
    p <- ggplot(es |> mutate(label = lulc_code),
                aes(x = .data[[pr[1]]], y = .data[[pr[2]]], label = label)) +
      geom_point(size = 3, colour = "#2166ac") +
      ggrepel::geom_text_repel(size = 2.5, max.overlaps = 10) +
      labs(x = pr[1], y = pr[2],
           title = paste("Trade-off:", pr[1], "vs", pr[2])) +
      theme_bw()
    fname <- paste0("scatter_", pr[1], "_vs_", pr[2], ".png")
    ggsave(file.path(output_dir, fname), p, width = 5, height = 4, dpi = 150)
    log_info("Saved: %s", fname)
  }, error = function(e) {
    log_error(
      "Failed to generate scatter plot for pair %s vs %s: %s\nProbable cause: column missing after filtering or ggrepel error.\nCheck: columns '%s' and '%s' exist and have valid data.\nPrevious skill: [none].",
      pr[1], pr[2], conditionMessage(e), pr[1], pr[2]
    )
    stop(e)
  })
}

log_info("Trade-off analysis completed. Outputs in: %s", output_dir)

File: skills/ecosystem-services-assessment/scripts/tradeoff_analysis.py
#!/usr/bin/env python3
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""
tradeoff_analysis.py
ES trade-off and synergy analysis across pixels or land cover units.
Usage: python tradeoff_analysis.py <es_summary_csv> <output_dir>
Requires: pandas, numpy, scipy, matplotlib, seaborn
"""
import logging
import sys
from datetime import datetime
from pathlib import Path
from itertools import combinations

SKILL_NAME = "ecosystem-services-assessment"
_LOG_DIR   = Path("logs")
_LOG_DIR.mkdir(parents=True, exist_ok=True)
_log_file  = _LOG_DIR / f"skill_{SKILL_NAME}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] [" + SKILL_NAME + "] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler(_log_file, encoding="utf-8"),
    ],
)
logger = logging.getLogger(SKILL_NAME)


def log_step(n: int, desc: str) -> None:
    logger.info("-- STEP %d: %s", n, desc)


def log_decision(var: str, val, why: str) -> None:
    logger.info("DECISION | %s = %s | %s", var, val, why)


import numpy as np
import pandas as pd
from scipy.stats import spearmanr
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import seaborn as sns


def main():
    # ── Arguments ────────────────────────────────────────────────────────────
    es_file    = sys.argv[1] if len(sys.argv) >= 2 else "outputs/ecosystem_services/es_summary_table.csv"
    output_dir = sys.argv[2] if len(sys.argv) >= 3 else "outputs/ecosystem_services"

    out_path = Path(output_dir)
    out_path.mkdir(parents=True, exist_ok=True)

    logger.info("Skill: %s | es_file=%s | output_dir=%s", SKILL_NAME, es_file, output_dir)

    # ── Input precondition check ─────────────────────────────────────────────
    if not Path(es_file).exists():
        logger.error(
            "Input not found: %s\n"
            "Probable cause: o script de quantificacao de servicos ecossistemicos nao foi executado ou o caminho esta errado.\n"
            "Check: execute primeiro o script de mapeamento/quantificacao de ES.\n"
            "Previous skill: ecosystem-services-assessment (quantification step).",
            es_file,
        )
        sys.exit(1)

    # ── Step 1: Load ─────────────────────────────────────────────────────────
    log_step(1, "Load ecosystem services table")
    try:
        es = pd.read_csv(es_file)
    except Exception as exc:
        logger.error(
            "Failed to read CSV de servicos ecossistemicos: %s\n"
            "Probable cause: corrupted file ou com separador incorreto.\n"
            "Check: abra o arquivo em editor de texto e confira o formato.\n"
            "Previous skill: ecosystem-services-assessment (quantification step).",
            exc,
        )
        sys.exit(1)

    logger.info("ES table loaded: %d land use classes", len(es))

    n_na_total = int(es.isna().sum().sum())
    if n_na_total > 0:
        logger.warning(
            "ES table contains %d NA values total — correlations will be computed with complete observations.",
            n_na_total,
        )

    # ── Step 2: Identify numeric ES columns ──────────────────────────────────
    log_step(2, "Identify ES indicator columns and normalise 0-1")
    exclude_cols = {"lulc_code", "n_pixels"}
    es_cols = [c for c in es.select_dtypes(include=[np.number]).columns if c not in exclude_cols]
    logger.info("ES indicators: %s", ", ".join(es_cols))
    log_decision(
        "es_cols", ", ".join(es_cols),
        "colunas numericas excluindo lulc_code e n_pixels sao tratadas como indicadores de ES",
    )

    if len(es_cols) < 2:
        logger.warning(
            "Fewer than 2 ES indicators found — trade-off analysis not possible with only %d column(s).",
            len(es_cols),
        )
        logger.info("Exiting without error. Add more indicators to the input CSV.")
        sys.exit(0)

    # ── Normalise to 0-1 ─────────────────────────────────────────────────────
    try:
        es_norm = es.copy()
        eps = 1e-10
        for col in es_cols:
            cmin = es_norm[col].min()
            cmax = es_norm[col].max()
            es_norm[col] = (es_norm[col] - cmin) / (cmax - cmin + eps)
    except Exception as exc:
        logger.error(
            "Failed to normalise indicators 0-1: %s\n"
            "Probable cause: non-numeric columns incorrectly identified as indicators.\n"
            "Check: column types in the input CSV.\n"
            "Previous skill: ecosystem-services-assessment (quantification step).",
            exc,
        )
        sys.exit(1)

    log_decision(
        "normalization", "min-max [0,1] with epsilon 1e-10",
        "evita divisao por zero quando todos os valores de um indicador sao iguais",
    )

    # ── Step 3: Correlation matrix (Spearman) ────────────────────────────────
    log_step(3, "Compute Spearman correlation matrix among ES indicators")
    try:
        subset = es_norm[es_cols].dropna()
        rho, _ = spearmanr(subset)
        if len(es_cols) == 2:
            # spearmanr returns scalar for 2 variables — reshape to matrix
            cor_mat = np.array([[1.0, rho], [rho, 1.0]])
        else:
            cor_mat = rho
        cor_df = pd.DataFrame(cor_mat, index=es_cols, columns=es_cols)
    except Exception as exc:
        logger.error(
            "Failed to compute correlation matrix: %s\n"
            "Probable cause: all values in some column are NA after normalisation.\n"
            "Check: presence of variation in ES indicators.\n"
            "Previous skill: ecosystem-services-assessment (quantification step).",
            exc,
        )
        sys.exit(1)

    log_decision(
        "correlation_method", "Spearman",
        "metodo nao-parametrico robusto a distribuicoes assimetricas comuns em dados de ES",
    )

    try:
        csv_path = out_path / "tradeoff_matrix.csv"
        cor_df.to_csv(csv_path)
        logger.info("tradeoff_matrix.csv saved in: %s", output_dir)
    except Exception as exc:
        logger.error(
            "Failed to salvar tradeoff_matrix.csv: %s\n"
            "Probable cause: permissao negada ou disco cheio.\n"
            "Check: permissoes do output directory.\n"
            "Previous skill: [none].",
            exc,
        )
        sys.exit(1)

    # ── Step 4: Correlation heatmap ──────────────────────────────────────────
    log_step(4, "Generate trade-off heatmap")
    try:
        fig, ax = plt.subplots(figsize=(8, 7))
        mask = np.triu(np.ones_like(cor_df, dtype=bool), k=1)
        sns.heatmap(
            cor_df, mask=mask, annot=True, fmt=".2f", cmap="RdBu_r",
            vmin=-1, vmax=1, center=0, square=True, linewidths=0.5, ax=ax,
        )
        ax.set_title("ES Trade-offs (Spearman r)")
        fig.tight_layout()
        fig.savefig(out_path / "tradeoff_heatmap.png", dpi=150)
        plt.close(fig)
        logger.info("tradeoff_heatmap.png saved in: %s", output_dir)
    except Exception as exc:
        logger.error(
            "Failed to generate heatmap de trade-offs: %s\n"
            "Probable cause: matplotlib/seaborn not installed ou matriz de correlacao invalida.\n"
            "Check: se os pacotes matplotlib e seaborn estao disponiveis e a matriz tem pelo menos 2 variaveis.\n"
            "Previous skill: [none].",
            exc,
        )
        sys.exit(1)

    # ── Step 5: Scatter plots for top pairs ──────────────────────────────────
    log_step(5, "Generate scatter plots for ES indicator pairs")
    pair_combos = list(combinations(es_cols, 2))
    n_pairs = min(6, len(pair_combos))
    logger.info("Generating %d scatter plots (of %d possible pairs)", n_pairs, len(pair_combos))
    log_decision(
        "max_scatter_plots", str(n_pairs),
        "limitado a 6 pares para evitar geracao excessiva de arquivos",
    )

    label_col = "lulc_code" if "lulc_code" in es.columns else None

    for v1, v2 in pair_combos[:n_pairs]:
        try:
            fig, ax = plt.subplots(figsize=(5, 4))
            ax.scatter(es[v1], es[v2], s=30, color="#2166ac", edgecolors="white", linewidths=0.5)
            if label_col is not None:
                for idx, row in es.iterrows():
                    ax.annotate(
                        str(row[label_col]),
                        (row[v1], row[v2]),
                        fontsize=7, alpha=0.7,
                        textcoords="offset points", xytext=(4, 4),
                    )
            ax.set_xlabel(v1)
            ax.set_ylabel(v2)
            ax.set_title(f"Trade-off: {v1} vs {v2}")
            fig.tight_layout()
            fname = f"scatter_{v1}_vs_{v2}.png"
            fig.savefig(out_path / fname, dpi=150)
            plt.close(fig)
            logger.info("Saved: %s", fname)
        except Exception as exc:
            logger.error(
                "Failed to generate scatter plot for pair %s vs %s: %s\n"
                "Probable cause: column missing after filtering or annotation error.\n"
                "Check: columns '%s' and '%s' exist and have valid data.\n"
                "Previous skill: [none].",
                v1, v2, exc, v1, v2,
            )
            sys.exit(1)

    logger.info("Trade-off analysis completed. Outputs in: %s", output_dir)


if __name__ == "__main__":
    main()
